### PR TITLE
feat: add EIP-8037 / TIP-1016 state gas support

### PR DESCRIPTION
**Pull request by rakita**
## Summary

Implements EIP-8037 (TIP-1016) dual-limit gas accounting, separating **execution gas** (CPU work) from **state gas** (storage/account creation, code deposit). This introduces a reservoir model where state gas draws from a separate pool before spilling into regular gas.

### Key Changes

- **Gas struct** (`interpreter/gas.rs`): Added `reservoir`, `regular_gas_budget`, `state_gas_spent` fields with `record_state_cost()` that draws from reservoir first, then spills to regular gas. `gas_left()` excludes reservoir per EIP-8037.
- **GasParams** (`context/interface/cfg/gas_params.rs`): New state gas cost IDs — `sstore_set_state_gas`, `new_account_state_gas`, `code_deposit_state_gas`, `create_state_gas`.
- **Charging sites** (5 locations):
  - `SSTORE` — new slot creation (`host.rs`)
  - `CREATE/CREATE2` — contract creation (`contract.rs`)
  - `Code deposit` — per-byte code storage (`frame.rs`)
  - `CALL` — new account with value transfer (`call_helpers.rs`)
  - `SELFDESTRUCT` — new beneficiary account (`host.rs`)
- **Frame propagation** (`handler/frame.rs`): Child frame state gas accumulates to parent on success; reservoir refill on halt/revert prevents state gas from permanently consuming regular gas budget.
- **Precompiles**: All precompile gas charges use `record_regular_cost()` — no state gas for precompiles.
- **GAS opcode** (`system.rs`): Returns `remaining - reservoir` per spec.
- **op-revm**: Updated to handle new gas propagation in `last_frame_result`.
- **Inspector/EIP-3155**: Logs state gas fields for tracing.

### Architecture

| Component | Details |
|-----------|---------|
| Gas struct | `remaining` = `gas_left` + `reservoir`; `regular_gas_budget` is implicit cap for regular gas |
| Regular gas | Checked against `gas_left` only via `record_cost()`, cannot draw from reservoir |
| State gas | Drawn from `reservoir` first via `record_state_cost()`, spills to `gas_left` when exhausted |
| GAS opcode | Returns `gas_left()` only, excluding the state gas reservoir |
| Child frames | Inherit parent's `regular_gas_budget` and `reservoir` directly |
| EIP-8037 | Higher gas costs for state-creating ops, auto-enabled for Amsterdam+ specs |
| TIP-1016 | Dual-limit model with separate state gas reservoir, opt-in via `CfgEnv` |

### Cfg Integration

- `Cfg::state_gas_ids()` returns `Option<GasParams>` — `None` disables state gas (backward compatible).
- `CfgEnv` stores state gas config behind `#[cfg(feature = "optional_state_gas_params")]`.

### Test Suite

35+ test vectors in `crates/ee-tests/src/eip8037.rs` covering 7 categories:

1. **SSTORE state gas** — new slot, overwrite, zero-to-zero, multiple slots
2. **CREATE/CREATE2 state gas** — empty code, with code, with SSTORE, large code, OOG
3. **CALL state gas** — new/existing account, with/without value, SELFDESTRUCT
4. **Regular gas cap enforcement** — OOG, sufficient, reservoir interaction, block gas limit
5. **Cross-frame propagation** — child success/revert/halt, nested 3+ levels
6. **Interactions** — refund, GAS opcode, spend_all, precompiles, state_gas_spent tracking
7. **Reservoir refill** — revert vs halt behavior, boundary cases

## Test Plan

- [x] `cargo build --workspace` passes
- [x] `cargo nextest run --workspace` — all tests pass
- [x] `cargo nextest run -p ee-tests` — all 35+ EIP-8037 tests pass
- [x] `cargo clippy --workspace --all-targets --all-features` — clean
- [x] `cargo fmt --all` — clean
- [ ] `no_std` compatibility check
- [ ] Ethereum state tests via `revme statetest`

### CODE REVIEW

File: crates/context/interface/src/cfg/gas_params.rs [hunk 1]
```diff
@@ -239,6 +239,12 @@ impl GasParams {
         table[GasId::tx_token_cost().as_usize()] = gas::STANDARD_TOKEN_COST;
         table[GasId::tx_base_stipend().as_usize()] = 21000;
 
+        // TIP-1016: State gas constants
+        table[GasId::sstore_set_state_gas().as_usize()] = 20000;
+        table[GasId::new_account_state_gas().as_usize()] = 25000;
+        table[GasId::code_deposit_state_gas().as_usize()] = 200;
+        table[GasId::create_state_gas().as_usize()] = 32000;
```
Comment (rakita) [inline crates/context/interface/src/cfg/gas_params.rs]: new GasParams configs

File: crates/context/interface/src/cfg/gas.rs [hunk 2]
```diff
@@ -112,7 +112,11 @@ pub const CALL_STIPEND: u64 = 2300;
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 pub struct InitialAndFloorGas {
     /// Initial gas for transaction.
-    pub initial_gas: u64,
+    pub initial_total_gas: u64,
```
Comment (rakita) [inline crates/context/interface/src/cfg/gas.rs:271]: Check if this is correct, or should be have initial_regular_gas and initial_state_gas. so total gas would be state+regular

File: crates/context/interface/src/cfg/gas.rs [hunk 3]
```diff
@@ -112,7 +112,11 @@ pub const CALL_STIPEND: u64 = 2300;
...
+    /// State gas component of initial_gas (subset of initial_gas).
```
Comment (rakita) [inline crates/context/interface/src/cfg/gas.rs]:
```suggestion
    /// State gas component of initial_gas (subset of initial_total_gas).
```

File: crates/context/interface/src/cfg/gas_params.rs [hunk 4]
```diff
@@ -315,6 +315,42 @@ impl GasParams {
...
+            table[GasId::create_state_gas().as_usize()] = 112 * CPSB;
+
+            // SSTORE refund for 0→X→0 restoration: state gas + regular gas
+            table[GasId::sstore_set_refund().as_usize()] = 32 * CPSB + 2800;
+
+            // EIP-7702 parameter updates under EIP-8037
+            // Total per auth charged pessimistically:
+            //   regular: PER_AUTH_BASE_COST_REGULAR (7500)
+            //   state: (PER_EMPTY_ACCOUNT 112 + PER_AUTH_BASE 23) × CPSB
+            table[GasId::tx_eip7702_per_empty_account_cost().as_usize()] = 7500 + (112 + 23) * CPSB;
```
Comment (rakita) [inline crates/context/interface/src/cfg/gas_params.rs:346]: Check if the value is correct,

File: crates/context/interface/src/cfg/gas_params.rs [hunk 5]
```diff
@@ -654,6 +722,39 @@ impl GasParams {
...
+    ///
+    /// The state gas portion reduces `initial_state_gas` directly (not subject to refund caps).
+    /// The regular gas portion goes through the standard 1/5 refund cap.
+    ///
+    /// # Returns
+    ///
+    /// `(state_refund, regular_refund)` for the given total refund.
+    #[inline]
+    pub fn split_eip7702_refund(&self, total_refund: u64) -> (u64, u64) {
+        let per_auth_refund = self.tx_eip7702_auth_refund();
```
Comment (rakita) [inline crates/context/interface/src/cfg/gas_params.rs:746]: Recheck this logic, it can probably be simplified

File: crates/inspector/src/gas.rs [hunk 6]
```diff
@@ -28,24 +30,42 @@ impl GasInspector {
...
         Self {
             gas_remaining: 0,
             last_gas_cost: 0,
+            state_gas_spent: 0,
+            reservoir: 0,
         }
     }
 
     /// Sets remaining gas to gas limit.
     #[inline]
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/inspector/src/gas.rs:56]: Ack, new fields added for reservoir

File: crates/context/src/cfg.rs [hunk 7]
```diff
@@ -243,11 +216,17 @@ impl<SPEC> CfgEnv<SPEC> {
...
+    /// Automatically enables EIP-8037 for AMSTERDAM and later.
     pub fn with_spec_and_mainnet_gas_params<OSPEC: Into<SpecId> + Clone>(
         self,
         spec: OSPEC,
     ) -> CfgEnv<OSPEC> {
-        self.with_spec_and_gas_params(spec.clone(), GasParams::new_spec(spec.into()))
+        let is_amsterdam = spec.clone().into().is_enabled_in(SpecId::AMSTERDAM);
+        let enable_amsterdam_eip8037 = self.enable_amsterdam_eip8037 || is_amsterdam;
+        let mut cfg = self.with_spec_and_gas_params(spec.clone(), GasParams::new_spec(spec.into()));
+        cfg.enable_amsterdam_eip8037 = enable_amsterdam_eip8037;
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/context/src/cfg.rs:228]: This is true, and something we discussed in slack. It is similar how `set_spec` didn't not change `gas_params`

File: crates/handler/src/precompile_provider.rs [hunk 8]
```diff
@@ -72,6 +72,56 @@ impl Clone for EthPrecompiles {
...
+        gas: Gas::new_with_regular_gas_and_reservoir(gas_limit, output.reservoir),
+        output: bytes,
+    };
+
+    // set state gas, reservoir is already set in the Gas constructor
+    result.gas.set_state_gas_spent(output.state_gas_used);
+
+    // spend used gas.
+    if output.status.is_success_or_revert() {
+        result.gas.record_regular_cost(output.gas_used);
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/handler/src/precompile_provider.rs]: Code is fine, but adding must_use to record_regular_cost would be nice touch

File: crates/precompile/src/interface.rs [hunk 9]
```diff
@@ -144,65 +267,117 @@ pub trait Crypto: Send + Sync + Debug {
...
+/// Calls a [`PrecompileEthFn`] and wraps the result into a [`PrecompileOutput`].
+///
+/// Use this in wrapper functions to adapt an eth precompile to the [`PrecompileFn`] signature:
+/// ```ignore
+/// fn my_precompile(input: &[u8], gas_limit: u64, reservoir: u64) -> PrecompileOutput {
+///     call_eth_precompile(my_eth_fn, input, gas_limit, reservoir)
+/// }
+/// ```
+#[inline]
+pub fn call_eth_precompile(
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/precompile/src/interface.rs:365]: Fixed

File: examples/custom_precompile_journal/src/precompile_provider.rs [hunk 10]
```diff
@@ -95,26 +95,15 @@ fn run_custom_precompile<CTX: ContextTr>(
...
-            };
-            let underflow = interpreter_result.gas.record_cost(output.gas_used);
-            if !underflow {
-                interpreter_result.result = InstructionResult::PrecompileOOG;
-            }
-            Ok(interpreter_result)
-        }
+        Ok(output) => Ok(InterpreterResult {
+            result: InstructionResult::Return,
+            gas: Gas::new(inputs.gas_limit),
```
[the remark this answers is not in the document]
Comment (rakita) [reply examples/custom_precompile_journal/src/precompile_provider.rs]: it is example, fixed

File: crates/handler/src/handler.rs [hunk 11]
```diff
@@ -268,12 +284,27 @@ pub trait Handler {
...
             ctx.cfg().spec().into(),
             ctx.cfg().is_eip7623_disabled(),
-        )
-        .map_err(From::from)
+        )?;
+
+        // EIP-8037: When state gas is enabled and gas_limit exceeds TX_MAX_GAS_LIMIT,
+        // the maximum gas_used = TX_MAX_GAS_LIMIT (since reservoir = gas_limit - cap
+        // is excluded from gas_used). Floor gas must not exceed this maximum.
+        if ctx.cfg().is_amsterdam_eip8037_enabled()
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/handler/src/handler.rs:297]: Good find. In EIP-8037 it is speced as `max(intrinsic_regular_gas, calldata_floor_gas_cost) > TX_MAX_GAS_LIMIT`
